include/alpaka/mem/buf/Traits.hpp

```diff
@@ -191,4 +191,82 @@ namespace alpaka
 
         ALPAKA_UNREACHABLE(allocBuf<TElem, TIdx>(host, extent));
     }
+
+    namespace detail
+    {
+        // TODO(bgruber): very crude
+        template<typename DevDst, typename DevSrc>
+        auto canZeroCopy(DevDst const& devDst, DevSrc const& devSrc) -> bool
+        {
+            if constexpr(std::is_same_v<DevDst, DevSrc>)
+                if(devSrc == devDst)
+                    return true;
+            return false;
+        }
+    } // namespace detail
+
+    //! Makes the content of the source view available on the device associated with the destination queue. If the
+    //! destination shares the same memory space as the source view, no copy is performed and the destination view is
+    //! updated to share the same buffer as the source view. Otherwise, a memcpy is performed from source to
+    //! destination view.
+    template<typename TQueue, typename TViewDst, typename TViewSrc>
+    ALPAKA_FN_HOST void makeAvailable(TQueue& queue, TViewDst& viewDst, TViewSrc const& viewSrc)
+    {
+        ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
+
+        if constexpr(std::is_same_v<TViewSrc, TViewDst>) // TODO(bgruber): lift this by converting buffer types
+            if(detail::canZeroCopy(getDev(viewDst), getDev(viewSrc)))
+            {
+#if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
+                std::cout << "zero_memcopy: copy elided\n";
+#endif
+                viewDst = viewSrc;
+                return;
+            }
+
+#if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
+        std::cout << "zero_memcopy: deep copy required\n";
+#endif
+        memcpy(queue, viewDst, viewSrc);
+    }
+
+    //! Makes the content of the source view available on the destination device. If the destination shares the same
+    //! memory space as the source view, no copy is performed and the source view is returned. Otherwise a newly
+    //! allocated buffer is created on the destination device and the content of the source view copied to it.
+    template<
+        typename TQueue,
+        typename TDevDst,
+        typename TViewSrc,
+        std::enable_if_t<isDevice<TDevDst>, int> = 0,
+        typename TViewDst = Buf<TDevDst, Elem<TViewSrc>, Dim<TViewSrc>, Idx<TViewSrc>>>
+    ALPAKA_FN_HOST auto makeAvailable(TQueue& queue, TDevDst const& dstDev, TViewSrc const& viewSrc) -> TViewDst
```

Comment:
Could you please extend the documentation and point to the fact that the user must take care that all operations on `viewSrc` are finished.
This was one point where I was always confused in this PR but now I understand that the user must link the queue to avoid data races.

```C++
alpaka::enqueue(quereWhereViewSrcIsUsed, event);
wait(destQueue, event);
auto newBuff = alpaka::makeAvailable(destQueue, destDev, viewSrc));
```

Reply:
@bernhardmgruber `dstDev` is available in the `queue` parameter, is it really required to have this extra parameter `queue`?

Reply:
> Could you please extend the documentation and point to the fact that the user must take care that all operations on viewSrc are finished.

If you really insist, I can do it. It is also not documented for `alpaka::memcpy` btw. and I think it is kind of obvious. It's also not true. You can have pending operations on the source view:

```c++
auto bufHost = alpaka::allocBuf<int, Idx>(devHost, extents);
auto bufAcc = alpaka::makeAvailable(queue, devAcc, bufHost);
alpaka::exec<Acc>(queue, workdiv, Kernel{}, alpaka::getPtrNative(bufAcc));
alpaka::makeAvailable(queue, bufHost, bufAcc); // pending kernel execution on bufAcc
```